src/planning/src/planning/velocity_planning.cpp

```diff
@@ -0,0 +1,87 @@
+#include "planning/velocity_planning.hpp"
+
+double VelocityPlanning::find_circle_center(PathPoint &point1, PathPoint &point2,
+                                            PathPoint &point3) {
+  double x1 = point1.position.x;
+  double y1 = point1.position.y;
+  double x2 = point2.position.x;
+  double y2 = point2.position.y;
+  double x3 = point3.position.x;
+  double y3 = point3.position.y;
+
+  PathPoint mid1 = PathPoint((x1 + x2) / 2, (y1 + y2) / 2, 0);
+  PathPoint mid2 = PathPoint((x2 + x3) / 2, (y2 + y3) / 2, 0);
+  double slope1 = (x2 != x1) ? ((y2 - y1) / (x2 - x1)) : MAXFLOAT;
+  double slope2 = (x3 != x2) ? ((y3 - y2) / (x3 - x2)) : MAXFLOAT;
+  double slope1_perpendicular = -1 / slope1;
+  double slope2_perpendicular = -1 / slope2;
+  double center_x = (slope1_perpendicular * mid1.position.x -
+                     slope2_perpendicular * mid2.position.x + mid2.position.y - mid1.position.y) /
+                    (slope1_perpendicular - slope2_perpendicular);
+  double center_y = slope1_perpendicular * (center_x - mid1.position.x) + mid1.position.y;
+  double radius = sqrt(pow(center_x - x2, 2) + pow(center_y - y2, 2));
+  return radius;
+}
+
+void VelocityPlanning::point_speed(std::vector<double> &radiuses, std::vector<double> &velocities) {
+  for (int i = 0; i < static_cast<int>(radiuses.size()); i++) {
+    double velocity = sqrt(abs(config_.braking_acceleration_ * radiuses[i]));
+    velocities.push_back(std::max(velocity, config_.safety_speed_));
+  }
+  velocities.back() = config_.safety_speed_;
+
+  return;
+}
+
+void VelocityPlanning::speed_limiter(std::vector<PathPoint> &points,
+                                     std::vector<double> &velocities) {
+  for (int i = static_cast<int>(points.size()) - 2; i >= 0; i--) {
+    double distance = 0;
+    double max_speed = velocities[i];
+
+    for (int j = i + 1; j < static_cast<int>(points.size())-1; j++) {
```

Comment:
Are we skipping the last point in the path?

Reply:
Yes, the last point's velocity is fixed to the minimum velocity, it's the starting point to calculate the velocity for the other points